server/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -213,29 +200,23 @@ public ScriptService(Settings settings, Map<String, ScriptEngine> engines, Map<S
             }
         }
 
-        int cacheMaxSize = SCRIPT_CACHE_SIZE_SETTING.get(settings);
-
-        CacheBuilder<CacheKey, Object> cacheBuilder = CacheBuilder.builder();
-        if (cacheMaxSize >= 0) {
-            cacheBuilder.setMaximumWeight(cacheMaxSize);
-        }
-
-        TimeValue cacheExpire = SCRIPT_CACHE_EXPIRE_SETTING.get(settings);
-        if (cacheExpire.getNanos() != 0) {
-            cacheBuilder.setExpireAfterAccess(cacheExpire);
-        }
-
-        logger.debug("using script cache with max_size [{}], expire [{}]", cacheMaxSize, cacheExpire);
-        this.cache = cacheBuilder.removalListener(new ScriptCacheRemovalListener()).build();
-
-        this.lastInlineCompileTime = System.nanoTime();
         this.setMaxSizeInBytes(SCRIPT_MAX_SIZE_IN_BYTES.get(settings));
-        this.setMaxCompilationRate(SCRIPT_MAX_COMPILATIONS_RATE.get(settings));
+        compiler = new ContextCompiler(
+                    SCRIPT_CACHE_SIZE_SETTING.get(settings),
+                    SCRIPT_CACHE_EXPIRE_SETTING.get(settings),
+                    compilationLimitsEnabled() ?
+                        SCRIPT_MAX_COMPILATIONS_RATE.get(settings):
+                        new Tuple<>(0, TimeValue.ZERO)
+                    );
+    }
+
+    boolean compilationLimitsEnabled() {
```

Comment:
Can you please add a comment explaining this is overridden for the test infrastructure?

Reply:
Added.